src/persist-client/src/internal/apply.rs

```diff
@@ -0,0 +1,436 @@
+        if &self.upper != state.upper() {
+            self.upper.clone_from(state.upper());
+        }
+        if &self.since != state.since() {
+            self.since.clone_from(state.since());
+        }
+    }
+}
+
+impl<K, V, T, D> Applier<K, V, T, D>
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
+{
+    pub(crate) async fn new(
+        cfg: PersistConfig,
+        shard_id: ShardId,
+        metrics: Arc<Metrics>,
+        state_versions: Arc<StateVersions>,
+        shared_states: &StateCache,
+    ) -> Result<Self, Box<CodecMismatch>> {
+        let shard_metrics = metrics.shards.shard(&shard_id);
+        let state = shared_states
+            .get(shard_id, || {
+                metrics
+                    .cmds
+                    .init_state
+                    .run_cmd(&shard_metrics, |_cas_mismatch_metric| {
+                        // No cas_mismatch retries because we just use the returned
+                        // state on a mismatch.
+                        state_versions.maybe_init_shard(&shard_metrics)
+                    })
+            })
+            .await?;
+        let cached_state = {
+            let state = state.lock().await;
+            CachedState {
+                shard_id,
+                seqno: state.seqno,
+                upper: state.upper().clone(),
+                since: state.since().clone(),
+            }
+        };
+        Ok(Applier {
+            cfg,
+            metrics,
+            shard_metrics,
+            state_versions,
+            state,
+            cached_state,
+        })
+    }
+
+    pub fn cached_state(&self) -> &CachedState<T> {
+        &self.cached_state
+    }
+
+    pub fn is_tombstone(&self) -> bool {
+        self.cached_state.upper.is_empty() && self.cached_state.since.is_empty()
+    }
+
+    #[cfg(test)]
+    pub async fn seqno_since(&self) -> SeqNo {
+        self.state.lock().await.seqno_since()
+    }
+
+    pub async fn all_fueled_merge_reqs(&self) -> Vec<FueledMergeReq<T>> {
+        // WIP safe to serialize this on the mutex?
+        self.state
+            .lock()
+            .await
+            .collections
+            .trace
+            .all_fueled_merge_reqs()
+    }
+
+    pub async fn snapshot(
+        &self,
+        as_of: &Antichain<T>,
+    ) -> Result<Result<Vec<HollowBatch<T>>, Upper<T>>, Since<T>> {
+        // WIP safe to serialize this on the mutex?
+        self.state.lock().await.snapshot(as_of)
+    }
+
+    pub async fn verify_listen(
+        &self,
+        as_of: &Antichain<T>,
+    ) -> Result<Result<(), Upper<T>>, Since<T>> {
+        // WIP safe to serialize this on the mutex?
+        self.state.lock().await.verify_listen(as_of)
+    }
+
+    pub async fn next_listen_batch(&self, frontier: &Antichain<T>) -> Option<HollowBatch<T>> {
+        // WIP safe to serialize this on the mutex?
+        self.state.lock().await.next_listen_batch(frontier)
```

Comment:
it seems like it's also possible that `cached_state` contains a full copy of the latest `State<K, V, T, D>`. that means we always keep 2 copies of State per shard per process, but would mean we could run all reads without any locking at all. offhand that doesn't seem like the right tradeoff, but I did want to make sure it was stated out loud somewhere :)

Reply:
something I probably didn't explain well is that there is one State per shard per process, but CachedState is the same one per Machine/Handle model as before. so if it was a full copy, we'd be back where we started

however, one thing that this does suggest to me though is two shared copies, one of which looks like I have it, and the other would only get the lock very briefly at the end of an apply_unbatched_cmd or fetch_and_update_state and be used only for snapshot/listen/etc and other read things. this could also maybe even replace CachedState. downsides are:
- we'd have to reason about two mutexes (potential deadlocks, potential additional latency if something surprising happens), I wonder if we could use something like AtomicPtr?
- 2 copies of state instead of one (seems fine IMO)
- complexity (unfortunate but probably fine)
- the second copy probably involves a lot of cloning (bummer but certainly better than stressing about the contention we're stressing about in the WIP question I have)

Reply:
ooh I see, didn't grok that CachedState is still per handle. the two shared copies is pretty interesting... it does seem very liberating to not worry about the impact of reads, but I'm not sure how that compares to reasoning about two mutexes 🤔

Reply:
I think I've gradually talked myself into the two mutex thing as this has sat, but we obviously want to make that change extremely carefully.

I'm also starting to worry about messing up rebases here, so as a result, I think I'd like to split the `state: State` -> `Arc<Mutex<State>>` bit into its own followup PR and leave everything else here to merge first. thoughts?

Reply:
sounds good, though would we need to worry about shipping just this PR in a release?

Reply:
in the "scary change" sense? commits 1+2+4 are meant to be as close to pure code movement as possible. the only real exception to that is the introduction of CachedState stuff, but I could easily pull that out too

Reply:
yes to the "scary change" sense, though I think I misunderstood the original question. I'm comfortable with commits 1+2+4 shipping first